test-data/unit/check-classes.test

```diff
@@ -1702,6 +1729,179 @@ class C:
+reveal_type(B() + C())   # E: Revealed type is 'builtins.int'
+
+[case testReverseOperatorWithOverloads2]
+from typing import overload, Union
+
+class Num1:
+    def __add__(self, other: Num1) -> Num1: ...
+    def __radd__(self, other: Num1) -> Num1: ...
+
+class Num2(Num1):
+    # TODO: This should not be an error. See https://github.com/python/mypy/issues/4985
+    @overload    # E: Signature of "__add__" incompatible with supertype "Num1"
+    def __add__(self, other: Num2) -> Num2: ...  
+    @overload
+    def __add__(self, other: Num1) -> Num2: ...
+    def __add__(self, other): pass
+
+    @overload
+    def __radd__(self, other: Num2) -> Num2: ...
+    @overload
+    def __radd__(self, other: Num1) -> Num2: ...
+    def __radd__(self, other): pass
+
+class Num3(Num1):
+    def __add__(self, other: Union[Num1, Num3]) -> Num3: ...
+    def __radd__(self, other: Union[Num1, Num3]) -> Num3: ...
+
+reveal_type(Num1() + Num2())  # E: Revealed type is '__main__.Num2'
+reveal_type(Num2() + Num1())  # E: Revealed type is '__main__.Num2'
+
+reveal_type(Num1() + Num3())  # E: Revealed type is '__main__.Num3'
+reveal_type(Num3() + Num1())  # E: Revealed type is '__main__.Num3'
+
+reveal_type(Num2() + Num3())  # E: Revealed type is '__main__.Num2'
+reveal_type(Num3() + Num2())  # E: Revealed type is '__main__.Num3'
+
+[case testDivReverseOperatorPython3]
+# No error: __div__ has no special meaning in Python 3
+class A1:
+    def __div__(self, x: B1) -> int: ...
+class B1:
+    def __rdiv__(self, x: A1) -> str: ...
+
+class A2:
+    def __truediv__(self, x: B2) -> int: ...
+class B2:
+    def __rtruediv__(self, x: A2) -> str: ...  # E: Signatures of "__rtruediv__" of "B2" and "__truediv__" of "A2" are unsafely overlapping
+
+A1() / B1()                 # E: Unsupported left operand type for / ("A1")
+reveal_type(A2() / B2())    # E: Revealed type is 'builtins.int'
+
+[case testDivReverseOperatorPython2]
+# flags: --python-version 2.7
+
+# Note: if 'from __future__ import division' is called, we use
+# __truediv__. Otherwise, we use __div__. So, we check both:
+class A1:
+    def __div__(self, x):
+        # type: (B1) -> int
+        pass
+class B1:
+    def __rdiv__(self, x):   # E: Signatures of "__rdiv__" of "B1" and "__div__" of "A1" are unsafely overlapping
+        # type: (A1) -> str
+        pass
+
+class A2:
+    def __truediv__(self, x):
+        # type: (B2) -> int
+        pass
+class B2:
+    def __rtruediv__(self, x):  # E: Signatures of "__rtruediv__" of "B2" and "__truediv__" of "A2" are unsafely overlapping
+        # type: (A2) -> str
+        pass
+
+# That said, mypy currently doesn't handle the actual division operation very
+# gracefully -- it doesn't correctly switch to using __truediv__ when
+# 'from __future__ import division' is included, it doesn't display a very
+# graceful error if __div__ is missing but __truediv__ is present...
+# Also see https://github.com/python/mypy/issues/2048
```

Comment:
Could you please add these details to the issue and raise its priority to at least normal? This future import is very important in loads of numeric code.